### ISSUE TITLE
google-genai [feature]: Context Caching

### ISSUE THREAD

**Issue by afirstenberg** [langchain-ai/langchain#23259]
### Privileged issue

- [X] I am a LangChain maintainer, or was asked directly by a LangChain maintainer to create an issue here.

### Issue Content

Gemini now allows a developer to create a context cache with the system instructions, contents, tools, and model information already set, and then reference this context as part of a standard query. It must be explicitly cached (ie - it is not automatic as part of a request or reply) and a cache expiration can be set (and later changed).

It does not appear to be supported in Vertex AI at this time.

Open issues:
* Best paradigm to add to cache or integrate with LangChain history system
* Best paradigm to reference

References:
* AI Studio / genai: https://ai.google.dev/gemini-api/docs/caching?lang=node
* REST: https://ai.google.dev/api/rest/v1beta/cachedContents#CachedContent

**Comment by abhiaagarwal** [comment]
I'd absolutely love this feature. I've currently hacked together my own `BaseChatModel` for this and it's heavily cut down on my costs.

From an API design perspective, it would make sense for there to be a cache parameter (or kwarg) on `System/HumanMessage` to signal to the ChatModel that it should cache said message.

**Comment by aperepel** [comment]
I'm reading the comments and getting a feeling you'd like to cache messages. IMO, the primary use case for this feature is to cache reference data, like larger docs, audio/video, or, maybe extensive elaborate system instructions text.

Looks like the caching is configured at LLM object creation time which is then used normally, would LangChain play a role at all here? Maybe there are other use cases where additional LangChain constructs could help streamline the experience.

And, as always, please think of GenAI and Vertex.AI in tandem when designing things, they are essentially the consumer and enterprise sides of the same AI.

Thanks!

**Comment by mhattingpete** [comment]
Any news regarding this feature? @afirstenberg

**Comment by KevinZJN** [comment]
Hello, we are students from the University of Toronto trying to contribute and are interested in this feature. Is this feature currently being implemented? We also have a few related questions:

1. Would you be able to provide any further details on this feature and possibly some use cases?
2. How do we expect the integration with Langchain history system to behave?

**Comment by bungrudi** [comment]
would love to contribute to this as well.

**Comment by afirstenberg** [comment]
I don't think it is stale - but I'm not sure there are any answers on the best way to handle this. It looks like every system has implemented this differently, so there is no good abstraction.

**Comment by karsrini1973** [comment]
We have a similar requirement as well.. For a subset of our use cases, we would like to cache common instructions as well as reference (knowledge) data in the model's context window and use them for multiple requests. Will be very interested in any approach that would enable us to make use of such model(specific) capabilities from within the Langchain framework.

**Comment by hongkongkiwi** [comment]
Still very relevant and is obviously an important feature as they are releasing it for Gemini 2.0 in a few days: https://ai.google.dev/gemini-api/docs/caching?lang=node

Previous model (1.5) gemini caching has been available for some time.

I can see this really relevant if you want to refer to a long conversation with images/video, or else a very long prompt.

**Comment by hongkongkiwi** [comment]
This feature is already supported with Anthropic in Langchain: https://js.langchain.com/docs/integrations/chat/anthropic/#prompt-caching

We already have a generic Langchain cache support: https://js.langchain.com/docs/how_to/llm_caching/

I would think it's a matter of having a plugin which supports the Google Caching API? The difference I think is that you cannot actually retrieve the cached content with google but instead you initialize the model with the cached content identifier and google will then simply append it.

**Comment by hongkongkiwi** [comment]
Confusingly it looks like this might already be implemented? https://github.com/langchain-ai/langchainjs/blob/993d0f8ce5861eb1eb4bcbadc45dd15f28ebc22f/libs/langchain-google-genai/src/chat_models.ts#L713

**Comment by vansh-pundir-rc** [comment]
I can help to create context caching implementation on langchain

**Comment by DineshThumma9** [comment]
is this issue closed are is work being done on some repo i  am interested  i want to contribute
